crates/next-custom-transforms/src/transforms/server_actions.rs

```diff
@@ -2882,3 +2803,113 @@ impl From<Name> for Box<Expr> {
     }
 }
+
+fn emit_error(error_kind: ServerActionsErrorKind) {
+    let (span, msg) = match error_kind {
+        ServerActionsErrorKind::ExportedSyncFunction {
+            span,
+            in_action_file,
+        } => (
+            span,
+            formatdoc! {
+                r#"
+                    Only async functions are allowed to be exported in a {directive} file.
+                "#,
+                directive = if in_action_file {
+                    "\"use server\""
+                } else {
+                    "\"use cache\""
+                }
+            },
+        ),
+        ServerActionsErrorKind::InlineUseCacheInClientComponent { span } => (
+            span,
+            formatdoc! {
+                r#"
+                    It is not allowed to define inline "use cache" annotated functions in Client Components.
+                    To use "use cache" functions in a Client Component, you can either export them from a separate file with "use cache" or "use server" at the top, or pass them down through props from a Server Component.
+                "#
+            },
+        ),
+        ServerActionsErrorKind::InlineUseServerInClientComponent { span } => (
+            span,
+            formatdoc! {
+                r#"
+                    It is not allowed to define inline "use server" annotated Server Actions in Client Components.
+                    To use Server Actions in a Client Component, you can either export them from a separate file with "use server" at the top, or pass them down through props from a Server Component.
+
+                    Read more: https://nextjs.org/docs/app/api-reference/functions/server-actions#with-client-components
+                "#
+            },
+        ),
+        ServerActionsErrorKind::InlineSyncFunction { span, is_action_fn } => (
+            span,
+            formatdoc! {
+                r#"
+                    {subject} must be async functions.
+                "#,
+                subject = if is_action_fn {
+                    "Server Actions"
+                } else {
+                    "\"use cache\" functions"
+                }
+            },
+        ),
+        ServerActionsErrorKind::MisplacedDirective {
+            span,
+            directive,
+            location,
+        } => (
+            span,
+            formatdoc! {
+                r#"
+                    The "{directive}" directive must be at the top of the {location}.
+                "#,
+                location = match location {
+                    DirectiveLocation::Module => "file",
+                    DirectiveLocation::FunctionBody => "function body",
+                }
+            },
+        ),
+        ServerActionsErrorKind::MisplacedWrappedDirective { span, directive } => (
+            span,
+            formatdoc! {
+                r#"
+                    The "{directive}" directive must be at the top of the file, and cannot be wrapped in parentheses.
```

Comment:
hmm probably unrelated to the refactor here, but shouldn't we have a variant of this for function bodies too?

Reply:
Yeah, I wondered the same. For some reason this use case was not included in #47069. I'm not sure under which circumstances a formatter would add the parentheses, and whether this would also be done for inline directives. Can follow-up if needed.

Reply:
Follow-up is included here: #73018.